src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LocalProxyVarsGen.java

```diff
@@ -0,0 +1,271 @@
+        }
+        localDeclarations.addAll(newBody);
+        ListBuffer<JCStatement> assigmentsBeforeSuper = new ListBuffer<>();
+        for (Symbol vsym : fieldToLocalMap.keySet()) {
+            Symbol local = fieldToLocalMap.get(vsym);
+            assigmentsBeforeSuper.append(make.at(constructor.pos()).Assignment(vsym, make.at(constructor.pos()).Ident(local)));
+        }
+        constructor.body.stats = localDeclarations.toList();
+        if (!assigmentsBeforeSuper.isEmpty()) {
+            JCTree.JCMethodInvocation constructorCall = TreeInfo.findConstructorCall(constructor);
+            if (constructorCall.args.isEmpty()) {
+                // this is just a super invocation with no arguments we can set the fields just before the invocation
+                // and let Gen do the rest
+                TreeInfo.mapSuperCalls(constructor.body, supercall -> make.Block(0, assigmentsBeforeSuper.toList().append(supercall)));
+            } else {
+                // we need to generate fresh local variables to catch the values of the arguments, then
+                // assign the proxy locals to the fields and finally invoke the super with the fresh local variables
+                int argPosition = 0;
+                ListBuffer<JCStatement> superArgsProxies = new ListBuffer<>();
+                for (JCExpression arg : constructorCall.args) {
+                    long flags = SYNTHETIC | FINAL;
+                    VarSymbol proxyForArgSym = new VarSymbol(flags, newLocalName("" + argPosition), types.erasure(arg.type), constructor.sym);
+                    JCVariableDecl proxyForArgDecl = make.at(constructor.pos).VarDef(proxyForArgSym, arg);
+                    superArgsProxies = superArgsProxies.append(proxyForArgDecl);
+                    argPosition++;
+                }
+                List<JCStatement> superArgsProxiesList = superArgsProxies.toList();
+                ListBuffer<JCExpression> newArgs = new ListBuffer<>();
+                for (JCStatement argProxy : superArgsProxies) {
+                    newArgs.add(make.at(argProxy.pos).Ident((JCVariableDecl) argProxy));
+                }
+                constructorCall.args = newArgs.toList();
+                TreeInfo.mapSuperCalls(constructor.body,
+                        supercall -> make.Block(0, superArgsProxiesList.appendList(assigmentsBeforeSuper.toList()).append(supercall)));
+            }
+        }
+    }
+
+    Name newLocalName(String name) {
+        return names.fromString("local" + target.syntheticNameChar() + name);
+    }
+
+    class FieldRewriter extends TreeTranslator {
+        JCMethodDecl md;
+        Map<Symbol, Symbol> fieldToLocalMap;
+
+        public FieldRewriter(JCMethodDecl md, Map<Symbol, Symbol> fieldToLocalMap) {
+            this.md = md;
+            this.fieldToLocalMap = fieldToLocalMap;
+        }
+
+        @Override
+        public void visitIdent(JCTree.JCIdent tree) {
+            if (fieldToLocalMap.get(tree.sym) != null) {
```

Comment:
How do we tell this not to rewrite a field access if there's no need (e.g. not in early context) ?